src/log-store/src/raft_engine/log_store.rs

```diff
@@ -0,0 +1,519 @@
+                        }
+                    }
+                    Err(e) => {
+                        let _ = tx.send(Err(e)).await;
+                        break;
+                    }
+                }
+
+                if start_idx > last_index {
+                    break;
+                }
+            }
+        });
+
+        let s = stream!({
+            while let Some(res) = rx.recv().await {
+                yield res;
+            }
+        });
+        Ok(Box::pin(s))
+    }
+
+    async fn create_namespace(&mut self, ns: &Self::Namespace) -> Result<(), Self::Error> {
+        let key = format!("{}{}", NAMESPACE_PREFIX, ns.id).as_bytes().to_vec();
+        let mut batch = LogBatch::with_capacity(1);
+        batch
+            .put_message::<Namespace>(SYSTEM_NAMESPACE, key, ns)
+            .context(RaftEngineSnafu)?;
+        self.engine
+            .read()
+            .await
+            .as_ref()
+            .context(IllegalStateSnafu)?
+            .write(&mut batch, true)
+            .context(RaftEngineSnafu)?;
+        Ok(())
+    }
+
+    async fn delete_namespace(&mut self, ns: &Self::Namespace) -> Result<(), Self::Error> {
+        let key = format!("{}{}", NAMESPACE_PREFIX, ns.id).as_bytes().to_vec();
+        let mut batch = LogBatch::with_capacity(1);
+        batch.delete(SYSTEM_NAMESPACE, key);
+        self.engine
+            .read()
+            .await
+            .as_ref()
+            .context(IllegalStateSnafu)?
+            .write(&mut batch, true)
+            .context(RaftEngineSnafu)?;
+        Ok(())
+    }
+
+    async fn list_namespaces(&self) -> Result<Vec<Self::Namespace>, Self::Error> {
+        let mut namespaces: Vec<Namespace> = vec![];
+        self.engine
+            .read()
+            .await
+            .as_ref()
+            .context(IllegalStateSnafu)?
+            .scan_messages::<Namespace, _>(
+                SYSTEM_NAMESPACE,
+                Some(NAMESPACE_PREFIX.as_bytes()),
+                None,
+                false,
+                |_k, v| {
+                    namespaces.push(v);
+                    true
+                },
+            )
+            .context(RaftEngineSnafu)?;
+        Ok(namespaces)
+    }
+
+    fn entry<D: AsRef<[u8]>>(&self, data: D, id: Id, ns: Self::Namespace) -> Self::Entry {
+        Entry {
+            id,
+            data: data.as_ref().to_vec(),
+            namespace_id: ns.id(),
+            ..Default::default()
+        }
+    }
+
+    fn namespace(&self, id: store_api::logstore::namespace::Id) -> Self::Namespace {
+        Namespace {
+            id,
+            ..Default::default()
+        }
+    }
+
+    async fn obsolete(&self, namespace: Self::Namespace, id: Id) -> Result<(), Self::Error> {
+        let obsoleted = self
+            .engine
+            .read()
+            .await
+            .as_ref()
+            .context(IllegalStateSnafu)?
+            .compact_to(namespace.id(), id);
```

Comment:
Looks like raft engine only deletes entries `< id`, but we expect `obsolete` to delete entries `<= id`. We might need to add a test for this.